gh-133143: Make information about the interpreter ABI more accessible (#137476)
python/cpython: Doc/library/sys.rst

always available. Unless explicitly noted otherwise, all variables are read-only.

.. data:: abi_info

   An object containing information about the ABI of the currently running
   Python interpreter.  The following attributes are available in cpython:

   *pointer_bits* is the width of pointers in bits, as an integer, equivalent
   to ``8 * sizeof(void *)``, i.e. usually ``32`` or ``64``.

>    *Py_GIL_DISABLED* is a boolean indicating whether the interpreter was built

Reviewer: I would prefer consistency between attribute names. For example, rename this one to `gil_disabled`.

I also dislike negative options. Would it make sense to rename the attribute to `gil_enabled`? If would be False in a Free Threaded build.

Reviewer 2: It could be `free_threaded`, to use the new “branding”.

Author: I'm open to the naming and will defer to your expertise, but I will say I am a little bit concerned about name proliferation. With the other flags [under suggestion](https://discuss.python.org/t/add-sys-abi-features-to-make-information-about-the-interpreter-abi-more-accessible/94422/6), integrating macro `Py_XXX_YYY` as `xxx_yyy` is straight-forward enough, but if we come up with a new name potentially for everything, it might get more opaque than necessary.
Of course, somehow the whole point of the exercise is largely to make existing information more easily accessible and understandable and clearer names, found with the benefit of hindsight, can play a role.

So, `gil_disabled` or `free_threaded`?

Reviewer: > So, gil_disabled or free_threaded?

free_threaded

Author: Done.

   with the GIL disabled, i.e. with the :option:`--disable-gil` option,
   aka free-threading.

>    *Py_DEBUG* is a boolean indicating whether the interpreter was built in

Reviewer: You may rename the attribute to `debug_build`.

Author: The above applies. Anyway, I would then lean towards simply `debug` in this case.

>    debug mode, i.e. with the :option:`--with-pydebug` option.

Reviewer: Can you add a link on "debug mode" to https://docs.python.org/dev/using/configure.html#python-debug-build ? Something like:

```
:ref:`debug mode <debug-build>`
```

Author: Done.

always available. Unless explicitly noted otherwise, all variables are read-only.

.. data:: abi_info

   A :term:`named tuple` holding information about the ABI of the interpreter.

   .. attribute:: abi_info.pointer_bits

      The width of pointers in bits, as an integer.

      * ``32``: 32-bit build
      * ``64``: 64-bit build

>       * ``None`` if this information is unknown or neither 32 nore 64 bits.

Reviewer 2: IMO, for different pointer sizes we should simply use different numbers.

To me, `None` would mean something like an ABI that doesn't use pointers at all. (In which case we don't have  a C-style ABI at all, and `abi_info` itself should be missing.)

always available. Unless explicitly noted otherwise, all variables are read-only.

.. data:: abi_info

   An object containing information about the ABI of the currently running
   Python interpreter.
   It should include information that affect the CPython ABI in ways that require
   a specific build of the interpreter chosen from variants that can co-exist on
   a single machine.  For example, it does not encode the base OS (Linux or
   Windows), but does include pointer size since some systems support both 32-
   and 64- builds.  The available entries are the same on all platforms; e.g.

>    pointer_size is available even on 64-bit-only architectures.

Reviewer 3: For new text, we can start sentence (fragments) on new lines, to reduce diff sizes in the future. Also insert the missing word 'bit'.
```suggestion
   It should include information that affect the CPython ABI in ways that
   require a specific build of the interpreter chosen from variants
   that can co-exist on a single machine.  
   For example, it does not encode the base OS (e.g. Linux or Windows),
   but does include pointer size since some systems support
   both 32- and 64-bit builds.
   The available entries are the same on all platforms;
   e.g. *pointer_size* is available even on 64-bit-only architectures.
```

   New entries should be added when needed for a supported platform, or (for
   enabling an unsupported one) by core dev consensus.  Entries should be removed
   following :pep:`387`.

> 

Reviewer 3: I don't think we should put "core dev consensus" in user-facing documentation, it's irrelevant to how the attribute should be used. I would move this statement to a comment in an appropriate location, or remove it.

Author: I moved it to the top of the `make_abi_info` function in the `sys` module.

>    The following attributes are available in cpython:

Reviewer: ```suggestion
   The following attributes are available in CPython:
```

>    The following attributes are available in CPython:

Reviewer 3: The lead para says "CPython ABI", so saying "in CPython" here is redundant.
```suggestion
   The following attributes are available:
```

   *pointer_bits* is the width of pointers in bits, as an integer, equivalent

>    to ``8 * sizeof(void *)``, i.e. usually ``32`` or ``64``.

Reviewer 3: Is it ever not 32 or 64 on supported platforms, or more generally any platform CPython builds on?

Reviewer 2: No, but if you do port CPython to a 128-bit platform, putting 128 here is explicitly the right thing to do.

   *free_threaded* is a boolean indicating whether the interpreter was built
   with the GIL disabled, i.e. with the :option:`--disable-gil` option,

>    aka free-threading.

Reviewer 3: having "i.e." and "aka" in the same sentence is clunky.
```suggestion
   *free_threaded* is a boolean indicating whether the supports running
   in free-threaded mode (i.e. with the GIL disabled).
   This reflects the presence of the :option:`--disable-gil` configure option.
```

   *debug* is a boolean indicating whether the interpreter was built in

>    :ref:`debug mode <debug-build>`, i.e. with the :option:`--with-pydebug` option.

Reviewer 3: On Windows, we don't use configure options. Splitting the sentences makes the text slightly more accurate. If you'd like, it might be nice to update each bullet for the Windows equivalent.
```suggestion
   *debug* is a boolean indicating whether the interpreter was built in
   :ref:`debug mode <debug-build>`.
   This reflects the presence of the :option:`--with-pydebug` configure option.
```

>    *byteorder* is a string indicating the native byte order, either ``'big'`` or

Reviewer: I suggest renaming the member to *byte_order* since there are two words: https://en.wiktionary.org/wiki/byte_order

By the way it seems like *endianness* word is more commonly used.

Reviewer 2: I'd keep the name used in [`sys.byteorder`](https://docs.python.org/3/library/sys.html#sys.byteorder) and [`int.from_bytes`](https://docs.python.org/3/library/stdtypes.html#int.from_bytes)/`to_bytes`.

Author: I'm reasonably happy with any of the three names. I have commented on the discussion thread to invite other opinions, but if that is ok for you, @vstinner, I would stick with `byteorder`?

Reviewer: > I'd keep the name used in sys.byteorder and int.from_bytes/to_bytes.

It makes sense, I'm fine with keeping `byteorder` name in this case.

Reviewer 3: This makes the fields a little inconsistent as we use underscores for the other attributes (good!), but I assume the trade-off is fine here.

>    ``'little'``.  This is the same as the :data:`byteorder` attribute.

Reviewer 3: Generally readers will think of the attribute with the module prefix.
```suggestion
   ``'little'``.
   This is the same as the :data:`sys.byteorder` attribute.
```

>    .. versionadded:: next

Reviewer 3: I would move this to the very top (below the `.. data` line) as it could get confusing in the future as we add new attributes, which will have their own `versionadded` directives.

always available. Unless explicitly noted otherwise, all variables are read-only.

.. data:: abi_info

   .. versionadded:: next

   An object containing information about the ABI of the currently running
   Python interpreter.
   It should include information that affect the CPython ABI in ways that
   require a specific build of the interpreter chosen from variants that can
   co-exist on a single machine.
   For example, it does not encode the base OS (Linux or Windows), but does
   include pointer size since some systems support both 32- and 64-bit builds.
   The available entries are the same on all platforms;

>    e.g. *pointer_size* is available even on 64-bit-only architectures.

Reviewer 4: _e.g._ would have been better spelled out as _for example_
